File: sdk/dotnet/CloudMock.cs
using System;
using System.Diagnostics;
using System.Net.Sockets;
using System.Net;
using System.Threading;

namespace CloudMock
{
    public class CloudMockServer : IDisposable
    {
        private Process _process;
        public int Port { get; }
        public string Endpoint => $"http://localhost:{Port}";
        public string Region { get; }

        public CloudMockServer(int port = 0, string region = "us-east-1", string profile = "minimal")
        {
            Port = port > 0 ? port : FindFreePort();
            Region = region;

            _process = new Process
            {
                StartInfo = new ProcessStartInfo
                {
                    FileName = "cloudmock",
                    Arguments = $"--port {Port}",
                    RedirectStandardOutput = true,
                    RedirectStandardError = true,
                    UseShellExecute = false,
                    CreateNoWindow = true,
                    Environment = {
                        ["CLOUDMOCK_PROFILE"] = profile,
                        ["CLOUDMOCK_IAM_MODE"] = "none"
                    }
                }
            };
            _process.Start();
            WaitForReady(TimeSpan.FromSeconds(30));
        }

        public void Dispose()
        {
            if (_process != null && !_process.HasExited)
            {
                _process.Kill();
                _process.WaitForExit(5000);
            }
            _process?.Dispose();
        }

        private void WaitForReady(TimeSpan timeout)
        {
            var deadline = DateTime.UtcNow + timeout;
            while (DateTime.UtcNow < deadline)
            {
                try
                {
                    using var client = new TcpClient();
                    client.Connect("127.0.0.1", Port);
                    return;
                }
                catch (SocketException)
                {
                    Thread.Sleep(100);
                }
            }
            throw new TimeoutException("CloudMock did not start in time");
        }

        private static int FindFreePort()
        {
            var listener = new TcpListener(IPAddress.Loopback, 0);
            listener.Start();
            int port = ((IPEndPoint)listener.LocalEndpoint).Port;
            listener.Stop();
            return port;
        }
    }
}

File: sdk/ruby/lib/cloudmock.rb
# frozen_string_literal: true

require "socket"
require "net/http"
require "uri"

module CloudMock
  class Server
    attr_reader :port, :endpoint

    def initialize(port: nil, region: "us-east-1", profile: "minimal")
      @port = port || find_free_port
      @region = region
      @profile = profile
      @endpoint = "http://localhost:#{@port}"
      @pid = nil
    end

    def start
      return self if @pid

      @pid = spawn(
        { "CLOUDMOCK_PROFILE" => @profile, "CLOUDMOCK_IAM_MODE" => "none" },
        "cloudmock", "--port", @port.to_s,
        out: File::NULL, err: File::NULL
      )

      at_exit { stop }
      wait_for_ready(30)
      self
    end

    def stop
      return unless @pid
      Process.kill("TERM", @pid) rescue nil
      Process.wait(@pid) rescue nil
      @pid = nil
    end

    def aws_config
      {
        endpoint: @endpoint,
        region: @region,
        credentials: Aws::Credentials.new("test", "test"),
        force_path_style: true
      }
    end

    private

    def find_free_port
      server = TCPServer.new("127.0.0.1", 0)
      port = server.addr[1]
      server.close
      port
    end

    def wait_for_ready(timeout)
      deadline = Time.now + timeout
      loop do
        raise "CloudMock did not start in time" if Time.now > deadline
        begin
          TCPSocket.new("127.0.0.1", @port).close
          return
        rescue Errno::ECONNREFUSED
          sleep 0.1
        end
      end
    end
  end

  def self.start(**opts)
    Server.new(**opts).start
  end
end
